ultravox/data/datasets.py

```diff
@@ -1044,6 +1048,12 @@ def _get_sample(self, row) -> VoiceSample:
             audio_transcript=transcript,
         )
 
+    def __len__(self):
+        try:
+            return len(self._dataset)
+        except:
+            return self.total_samples
+
```

Comment:
I think it would be better to ensure consistency between `config.total_samples` and `len(self._dataset)` within `__init__`.

- If `len(self._dataset)` is functional, `config.total_samples` should equal `len(self._dataset)`; otherwise, an exception should be raised so that we can catch it and update the experiment configuration file accordingly.
- If `len(self._dataset)` is not functional, we should set `self.total_samples = config.total_samples` and introduce a mechanism to count the actual number of samples iterated in the dataset. If the actual number (computed when a `StopIteration` exception is thrown) is less than `self.total_samples`, or if it exceeds `self.total_samples` at any point during iteration, we should print a warning message to avoid wasting compute time.

Reply:
What's the plan to make this work for datasets other than `GenericVoiceDataset`?

Reply:
I think in the near future, every dataset will either be able to be derived from GenericVoiceDataset, or it will inherit from it.  Also, other datasets, don't have the ability to use epochs right now, which is what implementing __len__ is primarliy for right now.

Reply:
But either way, I think my SizediterableDataset implmenation should work for other datasets as well